### PR TITLE
Optional royaltyInfo() for ERC721 and mint() fix

### PR DESCRIPTION
**Pull request by victor-ego**
- Implemented a royaltyInfo() under the [ERC-2981: NFT Royalty Standard](https://eips.ethereum.org/EIPS/eip-2981#optional-royalty-payments) method for marketplaces to send to NFT creators their corresponding royalties.
- Minor fix to the mint() method @notice comment and return
- Deleted unnecessary isMinter() and state var so that only the smart contract owner can mint NFTs.
- Included a constant for adding a royalty %

### CODE REVIEW

File: cookiecutter.json [hunk 1]
```diff
@@ -8,5 +8,7 @@
     "mintable": "y",
     "max_supply": "y",
     "max_supply_amount": 10000,
-    "permitable": "y"
+    "permitable": "y",
+    "royalties": "y",
+    "royalty_percentage": 10
```
Comment (fubuloubu) [inline cookiecutter.json:13]: If it's a percentage, probably better to do something like `salePrice * {{ percentage }} / 100` down below in the code

Comment (fubuloubu) [reply cookiecutter.json:13]: Can also use Decimals e.g.

```vyper
convert(convert(salePrice, decimal) * PERCENTAGE, uint256)
```

File: {{cookiecutter.project_name}}/contracts/NFT.vy [hunk 2]
```diff
@@ -89,8 +92,19 @@ event ApprovalForAll:
...
+# @param buyer of the NFT
+event RoyalityInfo:
+    fee: indexed(uint256)
+    transactionDate: indexed(uint256)
+    amount: uint256
+    buyer: address
+{%- endif %}
+
 owner: public(address)
-isMinter: public(HashMap[address, bool])
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/contracts/NFT.vy:93]: Why is this removed?

File: {{cookiecutter.project_name}}/contracts/NFT.vy [hunk 3]
```diff
@@ -275,6 +294,26 @@ def getApproved(tokenId: uint256) -> address:
...
+    //          is owed and to whom. Important; Not all marketplaces respect this, e.g. OpenSea
+    /// @param _tokenId - the NFT asset queried for royalty information
+    /// @param _salePrice - the sale price of the NFT asset specified by _tokenId
+    /// @return receiver - address of who should be sent the royalty payment
+    /// @return royaltyAmount - the royalty payment amount for _salePrice
+    """
+    # check creator of contract
+    assert msg.sender == self.owner
+    # log Payment
+    log RoyalityInfo(_salePrice / ROYALTY_PERCENTAGE, block.timestamp, _salePrice, msg.sender)
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/contracts/NFT.vy:313]: Can't log an event in a view method

File: {{cookiecutter.project_name}}/contracts/NFT.vy [hunk 4]
```diff
@@ -89,8 +92,19 @@ event ApprovalForAll:
...
+# @param fee price at which the owner as changed it
+# @param amount sold NFT
+# @param buyer of the NFT
+event RoyalityInfo:
+    fee: indexed(uint256)
+    transactionDate: indexed(uint256)
+    amount: uint256
+    buyer: address
+{%- endif %}
+
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/contracts/NFT.vy:106]:
```suggestion
```

File: {{cookiecutter.project_name}}/contracts/NFT.vy [hunk 5]
```diff
@@ -143,6 +157,11 @@ MAX_SUPPLY: constant(uint256) = {{cookiecutter.max_supply_amount}}
 
 {%- endif %}
 
+{%- if cookiecutter.royalties == 'y' %}
+# @dev Percentage of royalties for lifetime for the creator
+ROYALTY_PERCENTAGE: constant(uint256) = {{cookiecutter.royalty_percentage}}
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/contracts/NFT.vy:162]:
```suggestion
ROYALTY_PERCENTAGE: constant(decimal) = {{ cookiecutter.royalty_percentage / 100.0 }}
```

File: {{cookiecutter.project_name}}/contracts/NFT.vy [hunk 6]
```diff
@@ -275,6 +294,26 @@ def getApproved(tokenId: uint256) -> address:
...
+    """
+    /// @notice Called with the sale price to determine how much royalty
+    //          is owed and to whom. Important; Not all marketplaces respect this, e.g. OpenSea
+    /// @param _tokenId - the NFT asset queried for royalty information
+    /// @param _salePrice - the sale price of the NFT asset specified by _tokenId
+    /// @return receiver - address of who should be sent the royalty payment
+    /// @return royaltyAmount - the royalty payment amount for _salePrice
+    """
+    # check creator of contract
+    assert msg.sender == self.owner
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/contracts/NFT.vy:310]: This isn't needed (and in fact will block it's proper use)
```suggestion
```

File: {{cookiecutter.project_name}}/contracts/NFT.vy [hunk 7]
```diff
@@ -275,6 +294,26 @@ def getApproved(tokenId: uint256) -> address:
...
+    # log Payment
+    _royalty: uint256 = convert(_salePrice, decimal) * ROYALTY_PERCENTAGE, uint256 # Percentage that accepts decimals
+    log RoyalityInfo(_royalty, block.timestamp, _salePrice, msg.sender)
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/contracts/NFT.vy:313]: No logs are specified per the spec
```suggestion
```

File: {{cookiecutter.project_name}}/contracts/NFT.vy [hunk 8]
```diff
@@ -275,6 +294,26 @@ def getApproved(tokenId: uint256) -> address:
...
+    //          is owed and to whom. Important; Not all marketplaces respect this, e.g. OpenSea
+    /// @param _tokenId - the NFT asset queried for royalty information
+    /// @param _salePrice - the sale price of the NFT asset specified by _tokenId
+    /// @return receiver - address of who should be sent the royalty payment
+    /// @return royaltyAmount - the royalty payment amount for _salePrice
+    """
+    # check creator of contract
+    assert msg.sender == self.owner
+    # log Payment
+    _royalty: uint256 = convert(_salePrice, decimal) * ROYALTY_PERCENTAGE, uint256 # Percentage that accepts decimals
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/contracts/NFT.vy:312]:
```suggestion
    royalty: uint256 = convert(convert(_salePrice, decimal) * ROYALTY_PERCENTAGE, uint256)
```

File: {{cookiecutter.project_name}}/contracts/NFT.vy [hunk 9]
```diff
@@ -275,6 +294,26 @@ def getApproved(tokenId: uint256) -> address:
...
+    log RoyalityInfo(_royalty, block.timestamp, _salePrice, msg.sender)
+    return self.owner, _royalty
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/contracts/NFT.vy:314]:
```suggestion
    return self.owner, royalty
```

File: {{cookiecutter.project_name}}/contracts/NFT.vy (hunk 2 above)
Comment (fubuloubu) [reply {{cookiecutter.project_name}}/contracts/NFT.vy:93]: Undo this change please. If you want to propose this as a separate change, I would put it in a different PR

File: {{cookiecutter.project_name}}/contracts/NFT.vy [hunk 10]
```diff
@@ -494,25 +534,19 @@ def setApprovalForAll(operator: address, approved: bool):
     log ApprovalForAll(msg.sender, operator, approved)
 
 {%- if cookiecutter.mintable == 'y' %}
-@external
-def addMinter(minter: address):
-    assert msg.sender == self.owner
-    self.isMinter[minter] = True
-
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/contracts/NFT.vy:501]: Same as above, propose as a different PR

File: {{cookiecutter.project_name}}/contracts/NFT.vy [hunk 11]
```diff
@@ -494,25 +534,19 @@ def setApprovalForAll(operator: address, approved: bool):
...
+    @return bool confirming that the minting occurred 
     """
 {%- if cookiecutter.max_supply == 'y' %}
     assert MAX_SUPPLY > self.totalSupply
 {%- endif %} 
 
-    assert msg.sender == self.owner or self.isMinter[msg.sender], "Access is denied."
+    assert msg.sender == self.owner, "Only contract owner can mint"
 
     self.totalSupply += 1
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/contracts/NFT.vy:551]: This line should be after `self.balanceOf[receiver] += 1`, seems like a bug

**Comment by victor-ego** [comment]
@fubuloubu i think we got it now, thanks for the review!

**Review by fubuloubu — APPROVED** [review]
Just add a test case (when the template feature is active) that checks the percentage vs. the configuration of the template (use `pytest.approx`), something like this:

```py
expected_royalty = int({{ cookiecutter.percentage / 100.0 }} * ape.convert("1 ether", int))
assert nft.royaltyInfo(1, "1 ether") == pytest.approx(expected_royalty)
```

File: {{cookiecutter.project_name}}/contracts/NFT.vy (hunk 11 above)
Comment (fubuloubu) [reply {{cookiecutter.project_name}}/contracts/NFT.vy]: If you could open another PR to fix this issue, that'd be most appreciated!

File: {{cookiecutter.project_name}}/tests/conftest.py [hunk 12]
```diff
@@ -14,3 +14,7 @@ def receiver(accounts):
 @pytest.fixture(scope="session")
 def nft(owner, project):
     return owner.deploy(project.NFT)
+
+@pytest.approx()
+def royalty_percentage():
+    return 10.0
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/tests/conftest.py:20]: This isn't doing anything

**Review by fubuloubu — APPROVED** [review]

**Review by fubuloubu — APPROVED** [review]

**Review by fubuloubu — APPROVED** [review]

File: {{cookiecutter.project_name}}/tests/test_token.py [hunk 13]
```diff
@@ -120,4 +124,11 @@ def test_uri(nft, owner):
     assert nft.baseURI() == "new base uri"
     assert nft.tokenURI(1) == "new base uri/1"
     
-    {%- endif %}
\ No newline at end of file
+    {%- endif %}
+
+ {%- if cookiecutter.royalties == 'y' %}
+
+def test_royaltyInfo(nft, royalty_percentage):
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/tests/test_token.py:131]:
```suggestion
def test_royaltyInfo(nft):
```

File: {{cookiecutter.project_name}}/tests/test_token.py [hunk 14]
```diff
@@ -1,6 +1,5 @@
 import ape
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/tests/test_token.py:1]:
```suggestion
import ape
import pytest
```

File: {{cookiecutter.project_name}}/tests/test_token.py [hunk 15]
```diff
@@ -120,4 +124,11 @@ def test_uri(nft, owner):
...
     
-    {%- endif %}
\ No newline at end of file
+    {%- endif %}
+
+ {%- if cookiecutter.royalties == 'y' %}
+
+def test_royaltyInfo(nft, royalty_percentage):
+    expected_royalty = int({{ cookiecutter.royalty_percentage / 100.0 }} * ape.convert("1 ether", int))
+    assert nft.royaltyInfo(1, "1 ether") == ape.approx(expected_royalty)
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/tests/test_token.py:133]:
```suggestion
    assert nft.royaltyInfo(1, "1 ether") == pytest.approx(expected_royalty)
```

**Review by fubuloubu — APPROVED** [review]

File: {{cookiecutter.project_name}}/contracts/NFT.vy [hunk 16]
```diff
@@ -275,6 +283,23 @@ def getApproved(tokenId: uint256) -> address:
 
 ### TRANSFER FUNCTION HELPERS ###
 
+{%- if cookiecutter.permitable == 'y' %}
+### Royalty integration under the ERC-2981: NFT Royalty Standard
+@external
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/contracts/NFT.vy:288]:
```suggestion
@view
@external
```

**Review by fubuloubu — APPROVED** [review]

File: {{cookiecutter.project_name}}/tests/test_token.py [hunk 17]
```diff
@@ -120,4 +125,11 @@ def test_uri(nft, owner):
...
     
-    {%- endif %}
\ No newline at end of file
+    {%- endif %}
+
+ {%- if cookiecutter.royalties == 'y' %}
+
+def test_royaltyInfo(nft):
+    expected_royalty = int({{ cookiecutter.royalty_percentage / 100.0 }} * ape.convert("1 ether", int))
+    assert nft.royaltyInfo(1, "1 ether") == pytest.approx(expected_royalty)
```
Comment (fubuloubu) [inline {{cookiecutter.project_name}}/tests/test_token.py:134]:
```suggestion
    assert nft.royaltyInfo(1, "1 ether")[1] == pytest.approx(expected_royalty)
```

**Review by fubuloubu — APPROVED** [review]
